service/src/main/java/com/commercetools/pspadapter/payone/PaymentHandler.java

```diff
@@ -55,68 +52,90 @@ public PaymentHandler(String payoneInterfaceName, String tenantName,
         try {
-            for (int i = 0; i < RETRIES_LIMIT; i++) {
-                try {
-                    final PaymentWithCartLike paymentWithCartLike = commercetoolsQueryExecutor.getPaymentWithCartLike(paymentId);
-                    String paymentInterface = paymentWithCartLike.getPayment().getPaymentMethodInfo().getPaymentInterface();
-                    if (!payoneInterfaceName.equals(paymentInterface)) {
-                        logger.warn(tenantNameKeyValue, "Wrong payment interface name: expected [{}], found [{}]", payoneInterfaceName, paymentInterface);
-                        return new PaymentHandleResult(HttpStatusCode.BAD_REQUEST_400);
-                    }
-
-                    paymentDispatcher.dispatchPayment(paymentWithCartLike);
-                    return new PaymentHandleResult(HttpStatusCode.OK_200);
-                } catch (final ConcurrentModificationException cme) {
-                    logger.warn(tenantNameKeyValue, "Exception on dispatchPayment with id [{}]. Retry {} of {}", paymentId, i + 1, RETRIES_LIMIT);
-                    Thread.sleep(RETRY_DELAY);
-                }
-            }
-
-            logger.error(
-                tenantNameKeyValue,
-                "The payment [{}] couldn't be processed after {} retries", paymentId, RETRIES_LIMIT);
-            return new PaymentHandleResult(HttpStatusCode.ACCEPTED_202,
-                    format("The payment couldn't be processed after %s retries", RETRIES_LIMIT));
-
+            return processPaymentOrRetry(paymentId, 0);
+        } catch (final ConcurrentModificationException concurrentModificationException) {
+            return handleConcurrentModificationException(paymentId, concurrentModificationException);
         } catch (final NotFoundException | NoCartLikeFoundException e) {
             return handleNotFoundException(paymentId, e);
         } catch (final ErrorResponseException e) {
             return errorResponseHandler(e, paymentId);
-        } catch (final CompletionException e) {
-            return completionExceptionHandler(e, paymentId);
         } catch (final Exception e) {
-            return handleThrowableInResponse(e, paymentId);
+            return handleException(e, paymentId);
         }
     }
 
-     private PaymentHandleResult handleNotFoundException(@Nonnull String paymentId, @Nonnull Exception exception ) {
+    private PaymentHandleResult processPaymentOrRetry(
+        @Nonnull final String paymentId,
+        final int currentAttemptCount) throws InterruptedException {
+
+        try {
+            return processPayment(paymentId);
+        } catch (final ConcurrentModificationException concurrentModificationException) {
+            if (currentAttemptCount == RETRIES_LIMIT - 1) {
+                throw concurrentModificationException;
+            } else {
+                Thread.sleep(RETRY_DELAY);
+                return processPaymentOrRetry(paymentId, currentAttemptCount + 1);
```

Comment:
The rewrite to recursion makes code more elegant but introduces other issues as in general it consumes more memory than simple iteration (worst case scenario if it loops long enough it will run into StackOverflow error) and is slower.

Reply:
That is true but in this case it's only for the `RETRIES_LIMIT = 20` so its not significant and cannot lead to Stackoverflow. I would choose code readability in that case. WDYT?

Reply:
18485e5